cpp/dolfinx/fem/interpolate.h

```diff
@@ -835,103 +841,81 @@ void interpolate(Function<T>& u, std::span<const T> f,
-    using cmdspan2_t
-        = stdex::mdspan<const double, stdex::dextents<std::size_t, 2>>;
-    using mdspan2_t = stdex::mdspan<double, stdex::dextents<std::size_t, 2>>;
-    cmdspan2_t coords(coords_b.data(), 3, coords_b.size() / 3);
-    // Transpose interpolation coords
-    x.resize(coords.size());
-    mdspan2_t _x(x.data(), coords_b.size() / 3, 3);
-    for (std::size_t j = 0; j < coords.extent(1); ++j)
-      for (std::size_t i = 0; i < 3; ++i)
-        _x(j, i) = coords(i, j);
-  }
+  namespace stdex = std::experimental;
+  using cmdspan2_t
+      = stdex::mdspan<const double, stdex::dextents<std::size_t, 2>>;
+  using mdspan2_t = stdex::mdspan<double, stdex::dextents<std::size_t, 2>>;
+  cmdspan2_t coords(coords_b.data(), 3, coords_b.size() / 3);
+  // Transpose interpolation coords
+  x.resize(coords.size());
+  mdspan2_t _x(x.data(), coords_b.size() / 3, 3);
+  for (std::size_t j = 0; j < coords.extent(1); ++j)
+    for (std::size_t i = 0; i < 3; ++i)
+      _x(j, i) = coords(i, j);
 
   // Determine ownership of each point
-  auto [dest_ranks, src_ranks, received_points, evaluation_cells]
-      = geometry::determine_point_ownership(*mesh_v, x);
-
-  // Evaluate the interpolating function where possible
-  std::vector<T> send_values(received_points.size() / 3 * value_size);
-  v.eval(received_points, {received_points.size() / 3, (std::size_t)3},
-         evaluation_cells, send_values,
-         {received_points.size() / 3, (std::size_t)value_size});
-
-  // Send values back to owning process
-  std::array<std::size_t, 2> v_shape = {src_ranks.size(), value_size};
-  std::vector<T> values_b(dest_ranks.size() * value_size);
-  impl::scatter_values(comm, src_ranks, dest_ranks,
-                       std::span<const T>(send_values), v_shape,
-                       std::span<T>(values_b));
-
-  // Transpose received data
-  namespace stdex = std::experimental;
-  stdex::mdspan<const T, stdex::dextents<std::size_t, 2>> values(
-      values_b.data(), dest_ranks.size(), value_size);
-
-  std::vector<T> valuesT_b(value_size * dest_ranks.size());
-  stdex::mdspan<T, stdex::dextents<std::size_t, 2>> valuesT(
-      valuesT_b.data(), value_size, dest_ranks.size());
-  for (std::size_t i = 0; i < values.extent(0); ++i)
-    for (std::size_t j = 0; j < values.extent(1); ++j)
-      valuesT(j, i) = values(i, j);
-
-  // Call local interpolation operator
-  fem::interpolate<T>(u, std::span(valuesT_b.data(), valuesT_b.size()),
-                      {valuesT.extent(0), valuesT.extent(1)}, cells);
+  return dolfinx::geometry::determine_point_ownership(*mesh_v, x);
+}
+
+/// Generate data needed to interpolate discrete functions across different
+/// meshes. Interpolate on all cells in the mesh.
+///
+/// @param[out] u The function to interpolate into
+/// @param[in] v The function to interpolate from
+/// @tparam Scalar type
+template <typename T>
+M2MInterpolationData generate_m2m_interpolation_data(const Function<T>& u,
```

Comment:
The terminology is getting a bit confusing. We used `nonmatching_meshes`?

Reply:
Does the function actually use the Function u or v? If not, then it would be better to pass the FunctionSpaces.

Reply:
It doesn't use the functions directly but I think that passing the function makes "more sense" for the user as they are the objects he or she wants to work with. The function space is "an implementation detail".